op-batcher: simplify trySignal, reduce parameters

nonsense opened the pull request:
This PR is:
1. Removing `c` parameter and `publishSignal` argument.
2. Log a warning when the `publishSignal` channel is full, instead of silently dropping the signal.
3. Remove passing of `wg`, and handling it in the actors goroutines.
4. Removing passing of `receipts` channel.

nonsense committed:
op-batcher: simplify trySignal; log if channel is full

nonsense committed:
op-batcher: remove passing of waitgroup field

nonsense commented on op-batcher/batcher/driver.go:
```diff
@@ -302,7 +300,7 @@ func (l *BatchSubmitter) loadBlocksIntoState(ctx context.Context, start, end uin
+			l.tryPublishSignal(false)
 		}
 
 	}
```
Note that this provides little guarantee that we'll actually signal the publishing loop every 100 blocks. In tests, this channel is blocked, and we actually end up adding >1000 blocks (`Added L2 block to local state`), before the `publishSignal` is read/unblocked.

geoknee commented on op-batcher/batcher/driver.go:
```diff
@@ -173,21 +172,20 @@ func (l *BatchSubmitter) StartBatchSubmitting() error {
-	go l.blockLoadingLoop(l.shutdownCtx, l.wg, unsafeBytesUpdated, publishSignal) // sends on unsafeBytesUpdated (if throttling enabled), and publishSignal. Closes them both when done
+	go l.receiptsLoop(receiptsCh)                            // ranges over receiptsCh channel
+	go l.publishingLoop(l.killCtx, receiptsCh)               // ranges over publishSignal, spawns routines which send on receiptsCh. Closes receiptsCh when done.
+	go l.blockLoadingLoop(l.shutdownCtx, unsafeBytesUpdated) // sends on unsafeBytesUpdated (if throttling enabled), and publishSignal. Closes them both when done
```
I'm on board with the motivation of not having both:
1. `publishSignal` channel stored in state, _and_
2. passing it around between goroutines

Did you consider moving towards only having 2, instead of only having 1? 

There are advantages to 2 that I can think of:
a) you can easily inject a channel when unit testing the function (e.g. `l.blockLoadingLoop`) without needing to construct a full `BatchSubmitter`. 
b) it can help clarify ownership of that channel (here, `l.blockLoadingLoop` is responsible for closing it) and prevent other routines from accessing it when they shouldn't. This is particularly important when ensuring everything is cleaned up well on shutdown
c) generally avoiding bloating the state of `BatchSubmitter`

sebastianst replied:
I also feel that constructing the channel (or waitgroup) in a higher function to establish clear ownership and then passing it around is better function style. Avoids side effects from the channel being accessed in other contexts via the state. Even more so now that the receipts channel is still passed around but the publish channel isn't.

Or is it necessary to have some of those channels, and the waitgroup, in the batcher's state?

nonsense replied:
>Did you consider moving towards only having 2, instead of only having 1?

It is a required mechanism for communication, so it is 1 or 2. I prefer 1 because it is less arguments passed around. Furthermore these channels are required and are always the same, so 1 is more intuitive I think.

> a) you can easily inject a channel when unit testing the function (e.g. l.blockLoadingLoop) without needing to construct a full BatchSubmitter.

True, but seeing the diff between this change and `develop`, shows that we don't really do that.

> I also feel that constructing the channel (or waitgroup) in a higher function to establish clear ownership and then passing it around is better function style.

Agree on ownership, but passing a struct field doesn't make sense to me.

> Even more so now that the receipts channel is still passed around but the publish channel isn't.

Agree, we should probably unify these two.

> Or is it necessary to have some of those channels, and the waitgroup, in the batcher's state?

Not sure if it is `necessary`, but they are already there. When they are already fields, passing them as arguments doesn't make sense.

nonsense replied:
@sebastianst I also removed the passing of `receipts` channel, now all signalling/communication channels are fields, and no need to pass them around. They are central to the operation of the `batcher` and they really don't belong as arguments. Their lifecycle is governed by the entrypoint `StartBatchSubmitting` which I think makes sense, as we want to be able to start/stop the service multiple times during its lifecycle.

nonsense replied:
@geoknee , I think with this change we are not bloating the state of the `BatchSubmitter`, these are required channels, and as you said, we have two options - either pass them to all producers/consumers, or keep them in state (but certainly not both). I am opting for keeping a reference for them in the state, as they are central to its operation, and the code becomes more succinct (no need to propagate them across many function calls).

nonsense committed:
op-batcher: unify receipts channel with publishSignal, and move to a field

nonsense committed:
op-batcher: extract l.wg away from actor goroutines

nonsense commented:
@sebastianst @geoknee I updated the PR again with your suggestions. Please take a look when you have a chance.

1. I unified the `receipts` channel and the `publishSignal` channel - both are kept in state, and not passed as arguments, as they are central to the operation of the BatchSubmitter, and both are owned by `StartBatchSubmitting`

3. I extracted `l.wg` from the actor goroutines, again reducing unnecessary parameter passing, and `l.wg.Add` and `l.wg.Done` are in a lot closer proximity. Same for `innerWg` (which is now renamed).
